### PR TITLE
Replace address bytes conversions with strkey conversions.

### PR DESCRIPTION
**Pull request by dmkozh**
### What

Replace address bytes conversions with strkey conversions.

### Why

The existing conversions are pretty confusing due to account vs contract address dichotomy. Strkey is more natural format for cross-chain interaction, that is much less error-prone (at a minor performance cost).

More motivation in https://github.com/stellar/rs-soroban-sdk/issues/996

### Known limitations

I'm not 100% sure about metering here - I do charge the allocations, but there is also a couple for-loops (base32 encoding/decoding and CRC computation). These should be very cheap though.

### CODE REVIEW

File: soroban-env-host/src/host.rs [hunk 1]
```diff
@@ -2438,50 +2440,82 @@ impl VmCallerEnv for Host {
...
+                    ScErrorCode::InvalidInput,
+                    "unexpected strkey length",
+                    &[strkey_obj.to_val()],
+                ));
+            }
+            // Approximate the decoding cost as two vector allocations for the
+            // expected payload length (the strkey library does one extra copy).
+            Vec::<u8>::charge_bulk_init_cpy(PAYLOAD_LEN + PAYLOAD_LEN, self)?;
+            // Charge for the key copy to string.
+            // Vec::<u8>::charge_bulk_init_cpy(key.len() as u64, self)?;
```
Comment (jayz22) [inline soroban-env-host/src/host.rs:2492]: This this line supposed to be uncommented, or removed?

Comment (dmkozh) [reply soroban-env-host/src/host.rs]: Nice catch, I thought I can avoid a copy, but apparently no, so need to charge another copy.

**Review by jayz22 — APPROVED** [review]

**Review by leighmcculloch — APPROVED** [review]
Looks good. I didn't get to check the length calculations.

Some suggestions inline.

File: soroban-env-host/src/host.rs [hunk 2]
```diff
@@ -2437,50 +2439,82 @@ impl VmCallerEnv for Host {
...
-        let account_id = self.account_id_from_bytesobj(pk_bytes)?;
-        self.add_host_object(ScAddress::Account(account_id))
+        address: AddressObject,
+    ) -> Result<StringObject, Self::Error> {
+        let strkey = self.visit_obj(address, |addr: &ScAddress| {
+            // Approximate the strkey encoding cost with two vector allocations:
+            // one for the payload size (32-byte key/hash + 3 bytes for
+            // version/checksum) and  another one for the base32 encoding of
+            // the payload.
+            const PAYLOAD_LEN: u64 = 32 + 3;
```
[the remark this answers is not in the document]
Comment (jayz22) [reply soroban-env-host/src/host.rs:2452]: Yeah agree. Same for the converted length (`expected_key_len` defined below), which the strkey lib calls `canonical_len`.

[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-host/src/host.rs:2452]: I don't have a strong opinion on this, but for the strkey these would need to be much more involved. We're just lucky to only deal with 32-byte keys here.

File: soroban-env-common/env.json [hunk 3]
```diff
@@ -1964,54 +1964,30 @@
...
                 {
                     "export": "1",
-                    "name": "account_public_key_to_address",
+                    "name": "strkey_to_address",
                     "args": [
                         {
-                            "name": "pk_bytes",
-                            "type": "BytesObject"
+                            "name": "strkey",
+                            "type": "StringObject"
```
[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-common/env.json]: I've tried making this function polymorphic and accept either string or bytes. At the SDK layer we could provide two 'overloads' that take either string or bytes. 

I've thought about conversion fn a bit more, and I'm not sure if there is a safe way to do type conversion (that would require creating two differently typed objects pointing at the same memory, which I'm not sure if is supported).

File: soroban-env-common/env.json [hunk 4]
```diff
@@ -1964,54 +1964,30 @@
...
                         }
                     ],
                     "return": "AddressObject",
-                    "docs": "Converts a provided 32-byte Stellar account public key to the corresponding address. This is only useful in the context of cross-chain interoperability. Prefer directly using the Address objects whenever possible."
+                    "docs": "Converts a provided Stellar strkey address of an account or a contract ('G...' or 'C...' respectively) to an address object."
```
[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-common/env.json]: Added more documentation about the supported key types and also disclaimer to discourage using this arbitrarily.

File: soroban-env-common/env.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (leighmcculloch) [reply soroban-env-common/env.json]: Yeah I guess this falls into the annoying realm of how do you view strings, and Rust views strings in an opinionated and restrictive way.

I think it's also reasonable that someone may embed a non-base32 encoded strkey into a message and want to extract that also. We don't have to provide first-class functions to do all these things, but we need to provide the building blocks at least.

[the remark this answers is not in the document]
Comment (leighmcculloch) [reply soroban-env-common/env.json]: I think the hard requirements is at least being able to do:
- String to Address where string is the strkey
- Bytes to Address where the bytes is a string

In both cases we don't actually need UTF8, etc, a valid strkey is _only_ the characters of base32 (no padding) encoded in a single char per byte.

[the remark this answers is not in the document]
Comment (leighmcculloch) [reply soroban-env-common/env.json]: The advantage of a first class bytes to Address is we can write the rules on the limited set of conversion properties, vs if we make a generic bytes <> string conversion we have to figure out how encoding, etc works if our strings are backed by the Rust String (I hope they aren't, I haven't actually checked).

[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-common/env.json]: `String` is `&[u8]`, i.e. the same as `Bytes`. We don't enforce it to be utf8 and it's role is mostly semantical/informational.

[the remark this answers is not in the document]
Comment (leighmcculloch) [reply soroban-env-common/env.json]: Then all our problems disappear. A conversion should imo by cheap and a ref, not a copy?

[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-common/env.json]: > A conversion should imo by cheap and a ref, not a copy?

We don't support by-ref object conversions. An object is a reference to some `HostObject` enum on the host side and `HostObject` has a single variant (e.g. `String` or `Bytes`). So we can't just start interpreting `HostObject::String` as `HostObject::Bytes` or vice versa, not without a significant host object revamp which we definitely don't want to do now.

File: soroban-env-host/src/host.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (leighmcculloch) [reply soroban-env-host/src/host.rs:2452]: It's up to you folks if you want to embed and maintain this logic or move it into strkey. We can add functions to strkey to calculate the length values you need, but we don't have to.

**Review by leighmcculloch — APPROVED** [review]
👏🏻

File: soroban-env-host/src/host.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dmkozh) [reply soroban-env-host/src/host.rs:2452]: I would prefer to merge this PR without blocking on strkey changes. Given that it's a standard that's been around for a while, I don't envision having the constants in host as a large risk.

[the remark this answers is not in the document]
Comment (leighmcculloch) [reply soroban-env-host/src/host.rs:2452]: Agreed, there's no need to block on this.

**Pull request merged**